### PR TITLE
feat: KVBM improved external block matching

### PR DESCRIPTION
**Pull request by jthomson04**
@ziqif-nv 

## Summary by CodeRabbit

**New Features**
- Introduced a distributed Key-Value Block Manager (KVBM) with leader-worker architecture for scalable block management.
- Added advanced block transfer system supporting multiple storage types (device, host, disk) and locality-aware transfers (local, logical, remote).
- Implemented Python bindings and integration for vLLM KV cache management, including cache manager, request, and block list types.
- Added support for new block layouts, including per-layer separate storage and fully contiguous layouts.
- Enabled asynchronous and remote block transfer strategies, including CUDA, memcpy, and NIXL-based transfers.
- Provided detailed Rust and Python APIs for slot/block management, offloading, onboarding, and block lifecycle control.

**Bug Fixes**
- Improved error handling and validation throughout block transfer, offload, and onboarding workflows.
- Enhanced test coverage and robustness for distributed and local block management operations.

**Documentation**
- Added comprehensive documentation and test plans for block lifecycle, offload management, slot/block manager workflows, and distributed messaging.
- Updated and expanded Python type stubs and module-level documentation.

**Tests**
- Introduced extensive Rust and Python test suites covering distributed messaging, block transfer, slot/block management, and vLLM cache integration.
- Added parameterized and integration tests for new block layouts and distributed workflows.

**Chores**
- Updated development environment configurations for improved Python linting and container builds.
- Refactored and organized codebase for clarity, modularity, and extensibility.

### CODE REVIEW

**Comment by zhaohaidao** [comment]
@jthomson04 Hi,
I have a question. The changes only involve two files. Why does the coderabbitai summary involve so much content?

File: lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py [hunk 1]
```diff
@@ -117,19 +117,22 @@ def get_offloaded_computed_blocks(
...
         num_disk_computed_tokens = disk_block_count * self.block_size
 
-        num_external_hit_tokens = max(num_disk_computed_tokens, num_host_computed_tokens)
+        num_external_hit_tokens = num_host_computed_tokens + num_disk_computed_tokens
+
+        need_to_allocate = num_external_hit_tokens
 
-        need_to_allocate = num_external_hit_tokens - num_computed_tokens
 
         # In a full-prompt-hit case, we need to recompute the last token
```
Comment (tzulingk) [inline lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py:134]: just curious: Why do we need to recompute the last token if a full-prompt is hit?

File: lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py [hunk 2]
```diff
@@ -117,19 +117,22 @@ def get_offloaded_computed_blocks(
 
         sequence_hashes = self._create_slot(request)
 
-        host_owned_blocks, disk_owned_blocks = self.cache_manager.get_offloaded_computed_blocks(sequence_hashes)
+        remaining_sequence_hashes = sequence_hashes[num_computed_tokens // self.block_size:]
```
Comment (tzulingk) [inline lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py:120]: I am new to the team, do you know why my dynamo workspace doesn't have lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py?

```
tzulingk@66d8878-lcedt:~/workspace_venv/dynamo$ ls lib/bindings/python/src/dynamo/llm/
__init__.py  __pycache__/ tensorrtllm/ 
```

**Comment by jthomson04** [comment]
> @jthomson04 Hi, I have a question. The changes only involve two files. Why does the coderabbitai summary involve so much content?

Because when I created the MR, I accidentally set the target branch to main

File: lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py (hunk 2 above)
Comment (jthomson04) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py:120]: Are you on the right branch?

File: lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py (hunk 1 above)
Comment (ziqifan617) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py:134]: I learned from https://github.com/LMCache/LMCache/blob/dev/lmcache/integration/vllm/vllm_v1_adapter.py#L832

essentially, to generate the next token, we need not only the kv cache, but logits. so we need to let vllm re-computes the last token

File: lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py (hunk 2 above)
Comment (ziqifan617) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py:120]: it is because kvbm related python bindings like you pointed is not in main/release branch yet. only exist in our private branch. we will merge to main once ready

**Comment by zhaohaidao** [comment]
> > @jthomson04 Hi, I have a question. The changes only involve two files. Why does the coderabbitai summary involve so much content?
> 
> Because when I created the MR, I accidentally set the target branch to main

Got it. Thanks for the explanation

File: lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py (hunk 1 above)
Comment (tzulingk) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py:134]: Thanks for the explanation. Would you mind adding your explanation as part of the comment?

File: lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py (hunk 2 above)
Comment (tzulingk) [reply lib/bindings/python/src/dynamo/llm/vllm_integration/kv_cache_manager.py:120]: Thank you.
